base/base/LineReader.cpp

```diff
@@ -65,6 +65,27 @@ std::optional<LineReader::Suggest::WordsRange> LineReader::Suggest::getCompletio
         });
 }
 
+void LineReader::Suggest::addWords(const std::vector<std::string> & new_words)
+{
+    for (const auto & new_word : new_words)
+    {
+        if (std::binary_search(words.begin(), words.end(), new_word))
+            continue;
+
+        words.push_back(new_word);
+        words_no_case.push_back(new_word);
```

Comment:
For me it doesn't look reliable. You're going to modify the vector while other function of the `Suggest` structure returns iterators inside this vector. It doesn't look very safe. Also you don't call `updateSuggest()` until the main completion list has been loaded from the server and it doesn't sound very reasonable. What I suggest is to make `LineReader::Suggest` a well-rounded class with a mutex so it could add words from multiple threads safely:

```
class Suggest
{
public:
    using Words = std::vector<std::string>;

    /// Called by Suggest::load() and also by ClientBase::updateSuggest().
    void addWords(Words && new_words);
    
    /// We don't give away any internal iterators
    void getCompletions(Replxx::completions_t & completions, const String & prefix, size_t prefix_length) const;

private:
    Words words;
    Words words_no_case;
    std::mutex mutex;
    /// We don't need std::atomic<bool> ready{false} anymore!
};
```

Reply:
> For me it doesn't look reliable. You're going to modify the vector while other function of the Suggest structure returns iterators inside this vector. It doesn't look very safe.

getCompletions() returns iterator, but it is converted to `Replxx::completions_t` instantly

https://github.com/ClickHouse/ClickHouse/blob/1e8e21570ccb3b557588d25e191c969efcf2efd0/base/base/ReplxxLineReader.cpp#L182-L183

I haven't touched this to reduce the patch.
Will convert getCompletion to return vector over iterator range. 

And addWords should not be run in parallel with getCompletions, so this should be safe.

> Also you don't call updateSuggest() until the main completion list has been loaded from the server and it doesn't sound very reasonable.

Yes, otherwise Suggest should store newly added words separately, but I did not want to introduce extra complexity for this simple thing, since usually when user type something in the client suggestion is already loaded.

> What I suggest is to make LineReader::Suggest a well-rounded class with a mutex so it could add words from multiple threads safely:

I don't think that it requires lock, since like I said, addWords should not be called in parallel with getCompletions, but if you insist, I can go this way.

Reply:
Yes, I believe your code works, I mean it just doesn't look right. I mean using the variable `std::atomic<bool> ready` to synchonize stuff didn't look very nice even before your patch, and now you're adding complexity and it's going to make the things a bit less understandable. That's why I suggest to make one step further and make synchronization more obvious and easy to see.

Reply:
Ok, I've replaced that atomic with mutex.